replicate favicon

dffischer opened the pull request:
Newsreaders may request the icon as a symbol for the feed. Web browsers usually ask for it as well. Instead of producing an error in this case, this patch provides them with the possibly cached icon parsed from the vine homepage.

dffischer committed:
Replicate favicon

Newsreaders may request the icon as a symbol for the feed. As the user
image is too big, providing them with the icon parsed from the vine
homepage is appropriate for vine feeds.

tremby commented on routes/index.coffee:
```diff
@@ -2,5 +2,6 @@ express = require 'express'
 router = express.Router()
 
 router.get /^\/users\/(\d+)\/(atom|rss)$/, require './users'
+router.get /^.*\/favicon.ico$/, require './favicon'
```
Is this correct? This would match `anything-here/at-all/favicon.ico`, and besides would be more succinct as just `/\/favicon.ico$/`.

dffischer replied:
It _should_ be located in the root path. However, [as this article reports](https://perishablepress.com/block-favicon-url-404-requests/), there seem to be some entities that look for it deeper down. So I thought it could not hurt to satisfy them, too, as this will reduce the number of failed requests.

tremby replied:
OK, fair enough. So just change the regex to `/\/favicon.ico$/`, to more simply match anything ending in `/favicon.ico`.

dffischer replied:
Done.

tremby commented on routes/favicon.coffee:
```diff
@@ -0,0 +1,28 @@
+          error = new Error "Could not write favicon"
+          return next error
+        favicon req, res, next
+      
```
Please remove the trailing whitespace.

dffischer replied:
Done.

tremby commented on routes/favicon.coffee:
```diff
@@ -0,0 +1,28 @@
+cachefile = "favicon.ico"
+
+favicon = (req, res, next) ->
+  try
```
Please match the style of the other files. In particular here, tabs for indentation.

dffischer replied:
All leading white space was replaced with tabs.

tremby commented on routes/favicon.coffee:
```diff
@@ -0,0 +1,28 @@
+request = require 'request'
+fs = require 'fs'
+cachefile = "favicon.ico"
```
Please match the style of the other files. In particular here, single quotes for strings which are not one of text for humans or strings with `#{}` interpolation. (Admittedly this style probably wasn't obvious.)

dffischer replied:
I have few experience in coffee script, but this seems natural, coming from Python.

tremby commented on routes/favicon.coffee:
```diff
@@ -0,0 +1,28 @@
+      request (/https:\/\/[^"]*?favicon\.ico/.exec body)[0], {encoding: 'binary'}, (error, response, body) ->
+        return next error if error?
+        unless response.statusCode is 200
+          error = new Error "Unexpected response code #{response.statusCode} from while getting vine.co homepage"
```
This error message isn't correct.

Are you sure we even want an error here? Why not just act as a proxy and pass the response straight through?

dffischer replied:
I corrected the error messages. Having a dedicated error makes it easier to distinguish both cases, but I can change them to pass through instead, if you prefer that.

tremby commented on routes/favicon.coffee:
```diff
@@ -0,0 +1,28 @@
+				try
+					fs.writeFileSync cachefile, body, "binary"
+				catch error
+					error = new Error "Could not write favicon"
```
This should probably be a 500 too. (I can't actually remember what code this would return if not explicitly set, but let's be consistent with the rest of the code.)

dffischer replied:
Done.

tremby commented on routes/favicon.coffee:
```diff
@@ -0,0 +1,28 @@
+	try
+		res.send fs.readFileSync cachefile
+	catch error
+		request "https://vine.co/", (error, response, body) ->
```
Single quotes here please.

tremby replied:
Come to think of it, these read and write file calls should be using the async versions rather than the sync ones. Are you able to make that change? Let me know if you need some guidance.

While you're at it, can you make the `cachefile` variable in caps (to show it's meant as a constant) and separate it from the required modules by a line.

And where is that saving to? The current working directory, I'd assume, which could be anywhere the program is run from. I'm not sure this is a good idea. It'd probably be best to lock this to a specific place either alongside the vinefeed program itself (use `__dirname` which should give the `routes` directory this script is in, and amend from there) if you feel that makes sense, or a temporary directory elsewhere on the system (maybe try [os.tmpdir](https://nodejs.org/api/os.html#os_os_tmpdir)). I think the system temp dir makes more sense, but if you think alongside the program makes more sense please add the relevant path to the .gitignore file too.

dffischer replied:
Using only async functions where possible actually was my initial approach, but then I was deterred by the deep indentation. The implementation is now back to async again.

The icon was always saved to the directory the program runs in, which is its base directory since 0036188, unless someone changed the `CACHEFILE` to an absolute path. This was the intention in this constant. `os.tmpdir` is a step into the right direction, but more correct would be some cache directory. [The cachedir package](https://github.com/linusu/node-cachedir) only does the right thing when run as user, under root it should for example direct to `/var/cache/vinefeed`. I am still searching for a package that can universally find such a location, but used `os.tmpdir` until something pops up.

tremby replied:
All sounds very reasonable.

tremby replied:
> Using only async functions where possible actually was my initial approach, but then I was deterred by the deep indentation.

There are tricks to avoid this. I'm going to comment on the new patch with some suggestions.

tremby commented on routes/favicon.coffee:
```diff
@@ -0,0 +1,28 @@
+					error.status = 500
+					return next error
+				try
+					fs.writeFileSync cachefile, body, "binary"
```
Single quotes here please.

dffischer replied:
Done.

tremby commented on routes/favicon.coffee:
```diff
@@ -0,0 +1,28 @@
+
+favicon = (req, res, next) ->
+	try
+		res.send fs.readFileSync cachefile
```
I haven't tested, but does this send the correct mime type etc?

dffischer replied:
While using the async variants, I replaced this with the more direct `res.sendFile`. When accessed through Firefox, it shows up like any ordinary image would, so I trust the library on this one.

tremby commented on routes/favicon.coffee:
```diff
@@ -0,0 +1,28 @@
+				catch error
+					error = new Error "Could not write favicon"
+					return next error
+				favicon req, res, next
```
Could recursing here accidentally get us into an infinite loop / function call stack too deep error if writing succeeds but reading fails? I haven't thought through if this could actually happen or not, but I wonder if it could be avoided by sending the relevant response (passing through `response` itself?) here rather than recursing.

dffischer replied:
This could indeed have happened when the file was there but not readable. But I eliminated the recursive call while rewriting to the async function variants.

tremby commented on routes/favicon.coffee:
```diff
@@ -0,0 +1,27 @@
+request = require 'request'
+fs = require 'fs'
+tmpdir = require 'os'
+	.tmpdir()
```
You've just assigned a string to `tmpdir`. You're probably aware of that but I don't think an assignment like that should be with the module imports. I think instead you should do `os = require 'os'`, and then invoke `os.tmpdir()` later below.

tremby replied:
Oh, and personally if I _did_ want to write some code like what you have here I'd do it on one line, like `tmpdir = require('os').tmpdir()`.

dffischer replied:
This tried to save some function calls to the `tmpdir` function. But for the sake of consistence, I changed it to multiple function calls.

tremby commented on routes/favicon.coffee:
```diff
@@ -0,0 +1,27 @@
+	res.sendFile CACHEFILE,
+		root: tmpdir
+		(error) ->
+			if error
```
Since there's only one item in the object you're passing as options, I think it's neater to have it on the same line, and the signature of the callback too. It saves you a level of indentation, and looks clearer (at least to my eye). So you'd have this:

```
res.sendFile CACHEFILE, root: os.tmpdir(), (error) ->
    if error
```

tremby replied:
Then, notice you're doing a bunch of stuff `if error`, but nothing otherwise. You can skip another level of indentation by doing this the other way around.

```
return unless error
request 'htps://vine.co/', (error, response, body) ->
```

tremby replied:
I normally check for errors with the `?` existential operator too. `return unless error?`

dffischer replied:
Changed.

tremby commented:
I just realized we could instead just return a 301 or 302 redirection (open to discussion about which is more appropriate) straight to Vine's own favicon. I wonder if newsreader user agents would follow this. Do you have any idea?

dffischer commented:
A 301 seems more appropriate to me. It is a permanent redirection apart from moves that vine.co would make. I implemented this in [branch _facvicon-redirect_](https://github.com/dffischer/vinefeed/tree/favicon-redirect) and it seems to work fine with Firefox and Thunderbird.

As it is now only a url that can be cached in a variable, this saves all the hassle looking for a temporary or cache location. As even [stackexchange seems unable to find a cachedir](http://stackoverflow.com/questions/38573622/os-cachedir-find-cache-directory-in-node-js), I would prefer this approach.

Both branches were also rebased onto the new master. Should I create a new pull request for the branch?

tremby commented:
I'd say the redirection approach is the better one. Let's move to that! Thanks for your work here!

dffischer commented:
I incorporated all the comments and opened pull request #4.
